File: internal/storage/headwriter.go
package storage

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/micelio/micelio/internal/types"
)

// HeadResultWriter writes HEAD-only crawl results in JSONL or CSV format.
type HeadResultWriter struct {
	file   *os.File
	buf    *bufio.Writer
	csv    *csv.Writer
	format string
	count  int
}

// NewHeadResultWriter creates a writer for HEAD-only crawl output.
func NewHeadResultWriter(path string, format string) (*HeadResultWriter, error) {
	f, err := os.Create(path)
	if err != nil {
		return nil, fmt.Errorf("create output file: %w", err)
	}

	w := &HeadResultWriter{
		format: format,
		file:   f,
		buf:    bufio.NewWriterSize(f, 64*1024),
	}

	if format == "csv" {
		w.csv = csv.NewWriter(w.buf)
		if err := w.csv.Write(headCSVHeaders()); err != nil {
			f.Close()
			return nil, fmt.Errorf("write CSV header: %w", err)
		}
	}

	return w, nil
}

// Write writes a single HeadResult.
func (w *HeadResultWriter) Write(result *types.HeadResult) error {
	if w.format == "csv" {
		return w.writeCSV(result)
	}
	return w.writeJSONL(result)
}

// Count returns the number of results written.
func (w *HeadResultWriter) Count() int {
	return w.count
}

// Close flushes and closes the writer.
func (w *HeadResultWriter) Close() error {
	if w.csv != nil {
		w.csv.Flush()
	}
	if err := w.buf.Flush(); err != nil {
		w.file.Close()
		return err
	}
	return w.file.Close()
}

func (w *HeadResultWriter) writeJSONL(result *types.HeadResult) error {
	data, err := json.Marshal(result)
	if err != nil {
		return fmt.Errorf("marshal head result: %w", err)
	}
	if _, err := w.buf.Write(data); err != nil {
		return err
	}
	if err := w.buf.WriteByte('\n'); err != nil {
		return err
	}
	w.count++
	return nil
}

func (w *HeadResultWriter) writeCSV(result *types.HeadResult) error {
	if err := w.csv.Write(flattenHeadForCSV(result)); err != nil {
		return err
	}
	w.count++
	return nil
}

func headCSVHeaders() []string {
	return []string{
		"url", "final_url", "status_code",
		"redirect_chain_length", "redirect_chain",
		"response_time_ms", "content_type", "content_length",
		"server", "x_robots_tag", "link_canonical",
		"hsts", "csp", "x_frame_options",
		"referrer_policy", "cache_control", "error",
	}
}

func flattenHeadForCSV(r *types.HeadResult) []string {
	// Redirect chain
	chainLen := len(r.RedirectChain)
	chain := ""
	if chainLen > 0 {
		parts := make([]string, chainLen)
		for i, hop := range r.RedirectChain {
			parts[i] = fmt.Sprintf("%d:%s", hop.StatusCode, hop.URL)
		}
		chain = strings.Join(parts, " -> ")
	}

	contentLength := ""
	if r.ContentLength != nil {
		contentLength = fmt.Sprintf("%d", *r.ContentLength)
	}

	deref := func(s *string) string {
		if s == nil {
			return ""
		}
		return *s
	}

	return []string{
		r.URL,
		r.FinalURL,
		fmt.Sprintf("%d", r.StatusCode),
		fmt.Sprintf("%d", chainLen),
		chain,
		fmt.Sprintf("%d", r.ResponseTimeMs),
		r.ContentType,
		contentLength,
		r.Server,
		deref(r.XRobotsTag),
		deref(r.LinkCanonical),
		fmt.Sprintf("%t", r.HSTS),
		fmt.Sprintf("%t", r.CSP),
		deref(r.XFrameOptions),
		deref(r.ReferrerPolicy),
		deref(r.CacheControl),
		r.Error,
	}
}

File: src/head-writer.ts
import { createWriteStream, type WriteStream } from 'node:fs';
import { format } from 'fast-csv';
import type { HeadResult } from './types.js';

export class HeadResultWriter {
  private stream: WriteStream;
  private csvStream: ReturnType<typeof format> | null = null;
  private outputFormat: 'jsonl' | 'csv';

  constructor(outputPath: string, outputFormat: 'jsonl' | 'csv') {
    this.outputFormat = outputFormat;
    this.stream = createWriteStream(outputPath, { flags: 'w' });

    if (outputFormat === 'csv') {
      this.csvStream = format({ headers: true });
      this.csvStream.pipe(this.stream);
    }
  }

  write(result: HeadResult): void {
    if (this.outputFormat === 'jsonl') {
      this.stream.write(JSON.stringify(result) + '\n');
    } else if (this.csvStream) {
      this.csvStream.write(this.flattenForCsv(result));
    }
  }

  async close(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.stream.on('error', reject);
      if (this.csvStream) {
        this.csvStream.on('error', reject);
        this.csvStream.end();
        this.stream.on('finish', resolve);
      } else {
        this.stream.end(resolve);
      }
    });
  }

  private flattenForCsv(result: HeadResult): Record<string, string | number | boolean | null> {
    return {
      url: result.url,
      final_url: result.finalUrl,
      status_code: result.statusCode,
      redirect_chain_length: result.redirectChain.length,
      redirect_chain: result.redirectChain.map(h => `${h.statusCode}:${h.url}`).join(' -> ') || '',
      response_time_ms: result.responseTimeMs,
      content_type: result.contentType,
      content_length: result.contentLength ?? '',
      server: result.server,
      x_robots_tag: result.xRobotsTag || '',
      link_canonical: result.linkCanonical || '',
      hsts: result.hsts,
      csp: result.csp,
      x_frame_options: result.xFrameOptions || '',
      referrer_policy: result.referrerPolicy || '',
      cache_control: result.cacheControl || '',
      error: result.error || '',
    };
  }
}
